src/operator/batch_norm-inl.h

```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
+      const Dtype *data = in_data + off;
+      Dtype *odata = out_data + off;
+      for (size_t i = 0; i < matrixSize; ++i) {
+        onData(data++, odata++);
+      }
+    }
+  }
+
+  /*! \brief Fast-foreach when you don't care about the position other than channel */
+  template<typename Shape, typename OnData>
+  static inline void forEachFast(const Dtype *in_data, Dtype *out_data,
+                                 const Shape& shape, OnData onData) {
+    const size_t dim = shape.ndim();
+    const size_t num = shape[0];
+    const size_t channels = dim > 1 ? shape[1] : 1;
+    const size_t matrixSize = shape.Size() / (channels * num);
+
+    for (size_t batchItem = 0; batchItem < num; ++batchItem) {
+      #pragma omp parallel for
+      for (size_t channel = 0; channel < channels; ++channel) {
+        for (size_t i = 0; i < matrixSize; ++i) {
+          onData(channel, in_data++, out_data++);
+        }
+      }
+    }
+  }
+
+  /*! \brief Compute the mean of each input channel */
+  template<typename Shape>
+  static inline void computeMean(const Dtype *in_data,
+                                 const Shape &ishape,
+                                 const Shape &stride,
+                                 const Shape &oshape,
+                                 Dtype *save_mean) {
+    const size_t channelCount = ishape[1];
+
+    for (size_t i = 0, n = oshape.Size(); i < n; ++i) {
+      save_mean[i] = 0;
+    }
+
+    forEachFast(in_data, ishape,
+                [&save_mean](const size_t channel, const Dtype *in_data){
+                  save_mean[channel] += *in_data;
+                });
+
+    const size_t itemCount = ishape.Size() / channelCount;
+    for (size_t i = 0, n = channelCount; i < n; ++i) {
+      save_mean[i] /= itemCount;
+    }
+  }
+
+  static inline bool isWriting(const OpReqType ort) {
+    return ort == kWriteTo || ort == kWriteInplace;
+  }
+
+  /*! \brief inverse standard deviation <-> variance
+   * Note that these aren't entirely reversible due to eps
+   **/
+  #define VARIANCE_TO_INVSTD(__var$,    __eps$)   (1.0/sqrt((__var$) + Dtype(__eps$)))
+  #define INVSTD_TO_VARIANCE(__invstd$, __eps$)   ((1.0 / ((__invstd$) * (__invstd$))) - (__eps$))
+
+  /*! \brief Compute the variance of each input channel, as well as update moving mean/variants */
+  template<typename DType, typename  Shape>
+  inline void computeVariance(const DType *in_data,
+                              const Shape &ishape,
+                              const Shape &stride,
+                              const DType *mean_data,
+                              const Dtype eps,
+                              const Dtype momentum,
+                              const Shape &oshape,
+                              Dtype *save_std) {
+    for (size_t i = 0, n = oshape.Size(); i < n; ++i) {
+      save_std[i] = 0;
+    }
+    const size_t channelCount = ishape[1];
+    CHECK(oshape.Size() == channelCount);
+
+    forEachFast(in_data, ishape,
```

Comment:
How about using std accumulate for mean and variance?
And using std transform for converting data back?